crates/sui-graphql-rpc/src/types/coin.rs

```diff
@@ -290,34 +293,67 @@ impl Coin {
-
-                    query = query.filter(
-                        dsl::coin_type.eq(coin_type.to_canonical_string(/* with_prefix */ true)),
-                    );
-
-                    if let Some(owner) = &owner {
-                        // Leverage index on objects table
-                        query = query.filter(dsl::owner_type.eq(OwnerType::Address as i16));
-                        query = query.filter(dsl::owner_id.eq(owner.into_vec()));
-                    }
+        let mut conn: Connection<String, Coin> = Connection::new(false, false);
+
+        let checkpoint_viewed_at =
+            match validate_cursor_consistency(checkpoint_viewed_at, page.after(), page.before()) {
+                Ok(checkpoint_viewed_at) => checkpoint_viewed_at,
+                Err(e) => return Err(e),
+            };
+
+        let response = db
+            .execute_repeatable(move |conn| {
+                let (lhs, mut rhs) = Checkpoint::available_range(conn)?;
 
-                    query
-                })
+                if let Some(checkpoint_viewed_at) = checkpoint_viewed_at {
+                    if checkpoint_viewed_at < lhs || rhs < checkpoint_viewed_at {
+                        return Ok::<_, diesel::result::Error>(None);
+                    }
+                    rhs = checkpoint_viewed_at;
+                }
+
+                let result = page.paginate_raw_query::<StoredHistoryObject, _>(
+                    conn,
+                    move |element| {
+                        element.map(|obj| {
+                            let cursor = Cursor::new(HistoricalObjectCursor::new(
+                                obj.object_id.clone(),
+                                rhs as u64,
+                            ));
+                            cursor
+                        })
+                    },
+                    move || {
+                        let final_ = Object::build_raw_consistent_query(
+                            |query| Self::filter(query, coin_type.clone(), owner),
+                            lhs as i64,
+                            rhs as i64,
+                        );
+
+                        filter!(final_, "newer.object_version IS NULL")
+                    },
+                )?;
+
+                Ok(Some((result.0, result.1, result.2, rhs)))
             })
             .await?;
 
-        let mut conn = Connection::new(prev, next);
+        let Some((prev, next, results, rhs)) = response else {
+            return Ok(conn);
+        };
+
+        conn.has_previous_page = prev;
+        conn.has_next_page = next;
 
         for stored in results {
-            let cursor = stored.cursor().encode_cursor();
-            let object = Object::try_from_stored_object(stored, checkpoint_sequence_number)?;
+            let cursor = Cursor::new(HistoricalObjectCursor::new(
+                stored.object_id.clone(),
+                rhs as u64,
+            ))
+            .encode_cursor();
```

Comment:
I think the "consistent_cursor" API I suggested on the Object PR would help clean this up as well.